Official GF names by default

mrgriffin opened the pull request:
Use official GF type/species/move/item names by default, instead of the unofficial run-together names that have been the default until now. This is achieved by introducing some narrower fonts, allowing us to fit more characters into the same space; we dynamically scale down the font width as needed in UIs that display names (e.g. the Pokemon Summary Screen and Battle Interface).

**Species**
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/e296dc67-0c60-47ba-a12e-69d09eb50ebe) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/7ff0c4ef-0066-406b-8b19-7a4c374ba3ff) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/eb19d085-e6a5-416d-9ebc-263d9a0e439b) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/3f63d2e2-2908-414c-84cf-a40bf454a11c) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/cb7edd02-10d7-4bfd-8df9-a5ace4040791) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/b85b2201-5497-4a08-ba53-c5e5792926ae) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/0f96b902-c1a7-4c46-a98a-f695a74ac450) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/4a6c8531-1eb3-4415-8cd3-b4f4afddb5d3) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/2b906eb3-96a5-4035-808d-64d96cb68b96) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/9338cda2-6afb-454d-b8d9-07678455d649) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/4cb13646-59c6-4d7b-9d84-76f10b9638b5) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/16370042-d40a-46d0-8e58-af7f82eb1217) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/a3db00ad-e794-4657-829a-a2e466fbc477) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/9bdb21b1-1ad7-410d-bc97-0bb34f9a55a5) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/8e1f9d01-ec56-4e1f-984d-2a70869d19f0) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/c6d4cbac-1401-46c8-8029-a18fc43b7b75)

**Items**
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/1b1bf374-2c23-41af-8ce6-b6cb8a9fb5c7) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/7f6289f5-1309-48a1-9494-6b03e69db6e5) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/068cae0b-ebca-4386-99d3-c6428ee8d1dc) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/2bfdb76c-73b0-43eb-9c22-246b9fff445d) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/9c16bf81-1427-40e9-a66d-ca66ab4370bf) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/ca19bdb4-93d9-4581-bd69-b12a0f4b4c5e) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/81b8df31-5ab1-4a36-9958-87c35b838bc0) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/757bbcad-5992-42c9-a537-646d5b51ed59) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/a31550b3-97b5-4b5d-8894-7ccac2119e1c)

**Moves**
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/6609119f-f1b5-4c2d-bae4-cbab677f3f6f) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/b6cb9585-9750-4f7d-b129-155b3c5d00e1) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/c6d9ac8b-5476-46ab-84e5-ecf2df812d88) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/93e8d230-a92d-47b1-900d-4a671fadd03f) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/dcd62377-70f5-4ded-9ae6-2373ed74b78f)

**Abilities** (already expanded by default)
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/7e77879d-5b85-4eaf-8606-5dcb3b748f07)

**Types**
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/72eb1778-0463-43d4-80b3-f79385c9c502) ![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/5af23e4a-0831-423e-b6fe-c8d2892cf509)

Credits: @mrgriffin `FONT_NARROWER` tweaks, `FONT_SHORT_NARROW`/`FONT_SMALL_NARROWER` graphics, code; @ZnogyroP `FONT_NARROWER` graphics; @cfmnephrite idea.

Changelog:
- If the hack has changed fonts—e.g. to the FrLg ones—then they probably want to create their own `FONT_NARROWER`, `FONT_SMALL_NARROWER`, and `FONT_SHORT_NARROW`.
- If the hack has introduced more UIs that render names they probably want to use `GetFontIdToFit`—see this PR's changes for examples of that.
- If the hack has introduced names in list menus they can use `.fontIdMayNarrow = TRUE` on the list template.
- Types, species, moves, and items all have individual commits which enable support in the UI and turn on longer names by default. These can be reverted by downstream projects if they really want to.

Future Work:
- Think about how we want to handle 12-character nicknames in parts of the save without spare padding (maybe use `SaveBlock3`?).
- Consider disabling the unofficial names.
- Consider having official GF descriptions for moves/abilities where they would fit using a narrower font.

Closes #4249, #4248, #3915.

mrgriffin commented:
CF raised TM pocket in the bag and menus in PokéMarts that sell TMs as two other places to check before this can be un-drafted. If there are any other places, _please_ raise them. Move relearner? Move tutor?

Bassoonian commented:
I wonder if this needs a config - I can imagine this isn't within everyone's stilistical goals. I'd also argue that any character that exists in the normal font should exist in this font too; there's quite some people with translated hacks (I've definitely seen French, Spanish and German) that'd get understandably upset if their diacritics are missing here.

Some text messages with move names buffered in string vars may overflow, perhaps? Did we check contests? The HGSS dex (move lists, but also move-based evolutions)?

mrgriffin commented:
> CF raised TM pocket in the bag and menus in PokéMarts that sell TMs as two other places to check before this can be un-drafted. If there are any other places, _please_ raise them. Move relearner? Move tutor?

TMs in the mart work fine:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/03a11c18-02c3-4ef9-a0f8-e1d808244147)

TMs in the bag need dynamic font widths:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/906f6507-5efe-4f1b-ab75-6fb1b02cc9da)

Move relearner needs dynamic font widths:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/7da35f4f-762b-4475-bb1e-4f8cdbfd71fd)

Move tutor is fine, just goes via the normal learn screen:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/2c88f5d5-ec6d-4cba-85fc-e973cd26fc4e)

mrgriffin commented:
> I wonder if this needs a config - I can imagine this isn't within everyone's stilistical goals.

I don't think it should require a separate one from `B_EXPANDED_MOVE_NAMES`, unless somebody wants to argue that a user prefers their text clipping to it dynamically resizing. I feel like either users want it to fit, or want it to resize.

> I'd also argue that any character that exists in the normal font should exist in this font too; there's quite some people with translated hacks (I've definitely seen French, Spanish and German) that'd get understandably upset if their diacritics are missing here.

I think that's doable. From a quick glance over the font, I think the diacritics should mostly fit in 3px. Notable exceptions Ñ and whatever the R-looking thing is to the right of it.

> Some text messages with move names buffered in string vars may overflow, perhaps?

Possibly. Is there a good way to check for that?

>  Did we check contests?

No, and they'll definitely need dynamic resizing (same space as the battle UI).
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/6de73caa-d3f7-4166-8558-eb2e615254b2)

> The HGSS dex (move lists, but also move-based evolutions)?

Ughhh. How does this screen work again?

mrgriffin commented:
Just leaving a comment here: people may expect species names to get the same treatment. I think this would require a `FONT_SMALL_NARROWER` for the HP boxes, and is not something I currently plan on doing. It would be helpful to have a list of everywhere that a species name could be displayed.

If somebody else would like to tackle species names, they're welcome to contribute to this PR (or raise a separate PR) :)

LOuroboros commented:
I have a small idea which I guess is in line with Jasper's suggestion above.

How about adding a function that takes an original font label as a parameter *(Ex `GetNarrowFont(u8 fontId)`)* that would return either `fontId` or `FONT_NARROWER`, and would be used in places like `BattlePutTextOnWindow` and `PrintMoveOnWindow`?

A function that would allow a user to quickly disable the usage of the narrower font if they so chose to basically.

mrgriffin commented:
Why would anybody want to disable the narrower font, thus causing their text to clip?

I can imagine it being helpful to say "I'd rather the strings be shorter, so they don't exceed the box size" (which is what `B_EXPANDED_MOVE_NAMES` being off does), but I can't imagine it being helpful to say "I'd like the strings to exceed the box size".

I'm planning on having a function that would technically give a single place that lists all the fonts to test when dynamically narrowing fonts, so it's not that it would be hard to implement a config item... I just don't see how there's a use-case for having one.

I guess somebody else can come PR a config later if they really want to.

LOuroboros commented:
> I can imagine it being helpful to say "I'd rather the strings be shorter, so they don't exceed the box size"

It's exactly because of that.

I'm not going to sugarcoat it, this doesn't look good to me, I find it hard to read, so as a user I know that I would like to have a quick way to disable it:

![310026517-e40eb3ae-057c-4e54-8601-cdc4e2f8f0f5](https://github.com/rh-hideout/pokeemerald-expansion/assets/4485172/028ff104-c2a1-44ae-afc6-66ac6f0288c2)

> which is what `B_EXPANDED_MOVE_NAMES` being off does

Ah. If `B_EXPANDED_MOVE_NAMES == FALSE` disables this narrower font then please ignore what I said and carry on 👍

mrgriffin commented:
> Ah. If `B_EXPANDED_MOVE_NAMES == FALSE` disables this narrower font then please ignore what I said and carry on 👍

It doesn't exactly disable the narrower font, but because we use the widest font which will fit all the letters, the narrower font should never be used (unless there are shorter move names that currently clip, but I don't think that's the case?)

I think we're in agreement, but just for anybody else coming along—some of the letters being cut off looks like crap, so I don't think "the narrower font looks crap" is a good enough argument for a config item that chooses to cut letters off.

LOuroboros commented:
> > Ah. If `B_EXPANDED_MOVE_NAMES == FALSE` disables this narrower font then please ignore what I said and carry on 👍
> 
> It doesn't exactly disable the narrower font, but because we use the widest font which will fit all the letters, the narrower font should never be used (unless there are shorter move names that currently clip, but I don't think that's the case?)
> 
> I think we're in agreement, but just for anybody else coming along—some of the letters being cut off looks like crap, so I don't think "the narrower font looks crap" is a good enough argument for a config item that chooses to cut letters off.

I didn't express myself clearly enough and I also didn't understand the effect of the changes, so that was entirely my bad.

What I feared was to have that ugly Stomping Tantrum. Now, if the narrower font is only ever used based on the width of the total amount of glyphs meaning that compact move names would use the regular font as usual, then I have no complaints and my suggestion was pointless from the start. I apologize, it was my bad :)

mrgriffin commented:
> > Did we check contests?
>
> No, and they'll definitely need dynamic resizing (same space as the battle UI).

Turns out they actually have even less space! Closer to 59 pixels than the battle UIs 64. Clips even with the narrower font, and even with even more aggressive kerning\* on the `i` and `T` we'd still only save 2 pixels, meaning we'd blow the budget by a pixel.
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/fac8a49a-8601-4634-ab21-1dec214e5efb)

\* Which is probably an invasive implementation because it would mean altering `RenderText` and `GetStringWidth` to incorporate whatever kerning algorithm we settle on.

mrgriffin commented:
> TMs in the bag need dynamic font widths

Turns out these also are a pixel too narrow for Stomping Tantrum:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/19741fe9-f7ef-42f4-a02a-5fd0c65e922a)

mrgriffin commented:
Added support for contests and TM bag pockets. Have not yet added support for the move relearner because that uses a generic list menu and so the change is going to be a little more invasive.

If somebody did desperately want to reintroduce clipping they could stub out `GetFontIdToFit`, or they could alter `sNarrowerFontIds` to exclude the fonts they don't like (e.g. do `[FONT_NARROW] = -1,` to exclude `FONT_NARROWER`).

Note that there are failing tests because some move names _do_ clip in contests, e.g. First Impression (only the first failure is reported).

mrgriffin commented:
Added move relearner, and also the ability to do the same font narrowing on any list template with `.fontIdMayNarrow = TRUE` (only enabled for the move relearner, but users might want it for other things too).

![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/4ff65c03-e329-4f21-adb5-478866992af8)

The remaining questions are:
1. Does HGSS Pokédex need changing? Probably, but I haven't checked yet.
2. Are there any other places where move names are shown?
3. Are we okay with long names clipping in contests? Alex, Jasper, and myself think so.

mrgriffin commented:
> Does HGSS Pokédex need changing? Probably, but I haven't checked yet.

It mostly doesn't.
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/2bfaffe8-63c7-4294-a256-8d152f6ab094)
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/e3b9ad5d-e250-404d-8c37-393ec8ed96ed)

Very slight clipping on a 3rd stage evolution:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/567327ea-2fe8-4464-aad0-5ba05a571156)

cfmnephrite commented on graphics/fonts/latin_narrower.png:
Minor thing but would you be willing to remove the whitespace after "i"?

mrgriffin replied:
It's fine in the summary screen, but I found it really tricky to notice the `i` in things like `in` without that space, to my eyes the dot looks like a diacritic on the `n`:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/838573/aca84abc-9325-4803-84e2-8a23b7ea95af)

But other people can feel free to weigh in on that issue.

AlexOn1ine commented:
do we want this for 1.8 or 1.9?

mrgriffin commented:
> do we want this for 1.8 or 1.9?

If 1.9, might be worth having a `FONT_SMALLER_NARROWER` and rolling this change out to species names and item names.

I think it makes sense to let this cook for a release cycle on `upcoming`, just in case there are places that were missed. It doesn't seem as urgent now, due to other changes for 1.8.

mrgriffin commented:
Update: this is now targeting 1.9. My hope is to support official names for abilities, items, and species too.

mrgriffin commented:
Finally ready for review! I've updated the PR description with screenshots of (iirc) all the places that are affected by this PR. _Please_ note that although many of those screenshots only show narrow text, all names that fit with the vanilla font will continue using that font—it's only too-long names which are made narrower to fit.

Please keep the history when merging (and if any review comments are made, please give me the opportunity to rebase) :)

AsparagusEduardo commented on src/data/items.h:
```diff
@@ -2863,8 +2869,8 @@ const struct Item gItemsInfo[] =
-        .name = _("FosslzedFish"),
-        .pluralName = _("FosslzedFishes"),
+        .name = HANDLE_EXPANDED_ITEM_NAME("FosslzedFish", "Fossilized Fish"),
+        .pluralName = HANDLE_EXPANDED_ITEM_NAME("FosslzedFishes", "Fossilized Fishes"),
```
Isn't the plural to Fish also Fish?

mrgriffin replied:
I just copied what was already there! iirc, "fishes" means "many different species of fish", whereas plural "fish" is "many of the same species of fish". I'm not sure which is applicable here though.

AsparagusEduardo commented on include/config/battle.h:
```diff
@@ -257,4 +258,18 @@
+#define B_EXPANDED_ITEM_NAMES TRUE
+#undef POKEMON_NAME_LENGTH
+#define POKEMON_NAME_LENGTH 12
+#endif
```
I don't think it makes much sense for this to be here. Is there another place it can be moved to?

AsparagusEduardo commented on include/test/test.h:
```diff
@@ -215,6 +215,8 @@ static inline struct Benchmark BenchmarkStop(void)
 
 #define PARAMETRIZE if (gFunctionTestRunnerState->parameters++ == gFunctionTestRunnerState->runParameter)
 
+#define PARAMETRIZE_LABEL(f, label) if (gFunctionTestRunnerState->parameters++ == gFunctionTestRunnerState->runParameter && (MgbaPrintf_(":N%s: " f " (%d/%d)", gTestRunnerState.test->name, label, gFunctionTestRunnerState->runParameter + 1, gFunctionTestRunnerState->parameters), 1))
```
Shouldn't this have its own documentation entry like `PARAMETRIZE` does?

mrgriffin replied:
Weirdly, there's not any documentation for `include/test/test.h`-specific things, only for `include/test/battle.h`... This matters, because `PARAMETRIZE` is defined differently for each of them, and so `PARAMETRIZE_LABEL` won't work in battle tests atm.

AsparagusEduardo replied:
ah I see. Something to do properly in the future then.

AsparagusEduardo commented on src/battle_message.c:
```diff
@@ -3940,6 +3940,15 @@ void BattlePutTextOnWindow(const u8 *text, u8 windowId)
+        // a tile in pixels. We cannot check the actual width of the
+        // window because B_WIN_MOVE_NAME_1 and B_WIN_MOVE_NAME_3 are
+        // 16 wide for Z-move details.
+        printerTemplate.fontId = GetFontIdToFit(text, printerTemplate.fontId, printerTemplate.letterSpacing, 8 * 8);
```
Would `TILE_SIZE` make sense to be used here?

AsparagusEduardo commented on src/contest.c:
```diff
@@ -1636,7 +1636,9 @@ static void Task_ShowMoveSelectScreen(u8 taskId)
 
         FillWindowPixelBuffer(i + MOVE_WINDOWS_START, PIXEL_FILL(0));
-        Contest_PrintTextToBg0WindowAt(i + MOVE_WINDOWS_START, moveName, 5, 1, FONT_NARROW);
+	// 5 + 7*8 because there are 7 tiles and about 5 extra pixels
+	// available.
```
Spaces instead of tabs. Also, would `TILE_WIDTH` or similar make sense to be used here instead of 8?
```suggestion
        // 5 + 7*8 because there are 7 tiles and about 5 extra pixels available.
```

AsparagusEduardo commented on src/pokemon_summary_screen.c:
```diff
@@ -2795,12 +2795,28 @@ static void ResetWindows(void)
+
+static void PrintTextOnWindowToFit(u8 windowId, const u8 *string, u8 x, u8 y, u8 lineSpacing, u8 colorId)
+{
+    PrintTextOnWindowToFitPx(windowId, string, x, y, lineSpacing, colorId, gWindows[windowId].window.width * 8);
```
Should we use `TILE_WIDTH` or similar in these places?

AsparagusEduardo commented on test/text.c:
I would use defines added to the specific screen headers to be referenced here as well instead of hardcoding numbers.

AsparagusEduardo replied:
I think this is the only comment left to address 🤔

mrgriffin replied:
Ah, I have some rambling replies to this here: https://github.com/rh-hideout/pokeemerald-expansion/pull/4241#issuecomment-2041483107. I think I lean towards the current situation being the best of some bad options: computing the sizes as the width of a window minus a constant and then hardcoding what the result of that was in the tests. Ideally I'd share the same computation in the UI code and tests, but as a practical matter I think it would be messy to make the window templates and window IDs available in the header files :(

AsparagusEduardo replied:
oh my bad, missed your comment ^^;

I see what you mean. Maybe just having defines at the start of this file will help updating those values for users, then

mrgriffin replied:
Cool, so hopefully this is the last question: which ones do you think should be defines, and do you have suggestions for what to name them?

Many of them are now computed in terms of `WindowWidthPx`. Ones that aren't:
- HP bar nickname (54).
- Contest nicknames (50).
- TM/HM pocket item name (73).
- Berries pocket item name (73).
- Non-TM/HM, non-berries pocket item name (88).
- Move relearner move name (68).
- Party menu nicknames (50).
- Player PC item name (74).
- Pokédex species name (50).
- Mart item name (84).

AsparagusEduardo commented on test/text.c:
```diff
@@ -0,0 +1,358 @@
+    case ITEM_THUNDER_STONE:
+    case ITEM_GALARICA_WREATH:
+    case ITEM_STRAWBERRY_SWEET:
+    case ITEM_ELECTRIC_MEMORY:  // *
+    case ITEM_FIGHTING_MEMORY:  // *
+    case ITEM_GROUND_MEMORY:    // *
+    case ITEM_PSYCHIC_MEMORY:   // *
+    case ITEM_DRAGON_MEMORY:    // *
+    case ITEM_CHARIZARDITE_X:   // *
+    case ITEM_CHARIZARDITE_Y:   // *
+    case ITEM_ULTRANECROZIUM_Z: // *
+    case ITEM_DEEP_SEA_SCALE:   // *
+    case ITEM_DEEP_SEA_TOOTH:   // *
+    case ITEM_NEVER_MELT_ICE:   // *
+    case ITEM_WEAKNESS_POLICY:  // *
+    case ITEM_SAFETY_GOGGLES:   // *
+    case ITEM_ADRENALINE_ORB:   // *
+    case ITEM_TERRAIN_EXTENDER: // *
+    case ITEM_PROTECTIVE_PADS:  // *
+    case ITEM_HEAVY_DUTY_BOOTS: // *
+    case ITEM_UTILITY_UMBRELLA: // *
+    case ITEM_MARANGA_BERRY:    // *
+    case ITEM_PUNCHING_GLOVE:   // *
+    case ITEM_AUSPICIOUS_ARMOR:
+    case ITEM_BOOSTER_ENERGY:   // *
```
What are these about?

mrgriffin replied:
The stars are next to items that have hold effects (i.e. are likely to be held by a Pokémon). Doesn't really do anything, except if somebody wanted to address just those ones.

AsparagusEduardo replied:
I'm not sure it's relevant to this test. Even if they have no use when held, items would still be displayed in UIs and such.

mrgriffin replied:
Yeah, that's true.

I think there's some use in being able to say "these are the important items, so I'm going to change the UI so that at least these ones fit, but I don't mind if some of the items which don't have hold effects don't quite fit".

But I can take the `*`s away if we don't think it's useful :)

AsparagusEduardo replied:
Maybe they could be grouped like this then
```c
    case ITEM_BLACK_APRICORN:
    case ITEM_THUNDER_STONE:
    case ITEM_GALARICA_WREATH:
    case ITEM_STRAWBERRY_SWEET:
    // Items with hold effects.
    case ITEM_ELECTRIC_MEMORY:
    case ITEM_FIGHTING_MEMORY:
    case ITEM_GROUND_MEMORY:
    case ITEM_PSYCHIC_MEMORY:
```

mrgriffin replied:
Oh, that's way better than what I wrote!

AsparagusEduardo commented on test/text.c:
```diff
@@ -0,0 +1,358 @@
+        if (gItemsInfo[i].importance) continue;
+        PARAMETRIZE_LABEL("%S", gItemsInfo[i].name) { item = i; }
+    }
+    switch (item)
```
Can't this be a check based on item name length so it supports custom item names?

mrgriffin replied:
That wouldn't make sense—the point here is to list just the items that are too long.

AsparagusEduardo replied:
After a talk on Discord, I now get what you meant by this ^^;

AsparagusEduardo commented on src/pokedex.c:
```diff
@@ -2347,6 +2347,17 @@ static void PrintMonDexNumAndName(u8 windowId, u8 fontId, const u8 *str, u8 left
     AddTextPrinterParameterized4(windowId, fontId, left * 8, (top * 8) + 1, 0, 0, color, TEXT_SKIP_DRAW, str);
 }
 
+static void PrintMonName(u8 windowId, u8 fontId, const u8 *str, u8 left, u8 top)
```
I don't see the harm of keeping these in the same function to avoid duplication of code. Number will very likely never be narrowed, so it shouldn't be affected by this.

mrgriffin replied:
I think I prefer making it clear which strings are intended to be affected by narrowing. What if I cleaned them up into:
```c
    static const u8 color[3] = { TEXT_COLOR_TRANSPARENT, TEXT_DYNAMIC_COLOR_6, TEXT_COLOR_LIGHT_GRAY };
    fontId = GetFontIdToFit(str, fontId, 0, 50);
    AddTextPrinterParameterized4(windowId, fontId, left * 8, (top * 8) + 1, 0, 0, color, TEXT_SKIP_DRAW, str);
```
And that, but without the middle line. For me personally, that doesn't feel like duplication at all, it's just two similar lines which also appear all over the place in the rest of the code, because it takes two lines to set up a color array and use that to draw text.

AsparagusEduardo replied:
Gotcha 👍

AsparagusEduardo reviewed (changes requested):
First review (focused on the code itself without opening the game)
I would prefer that we add defines for the window widths of all uses of `GetFontIdToFit`, `PrintTextOnWindowToFitPx`, `GetFontIdToFit` and others across the source, instead of hardcoding numbers.
Also, please separate the multiplication factors by spaces (`X * Y` instead of `X*Y`) .

mrgriffin commented:
> I would prefer that we add defines for the window widths of all uses of GetFontIdToFit, PrintTextOnWindowToFitPx, GetFontIdToFit and others across the source, instead of hardcoding numbers.

idk about this one. It doesn't seem that helpful to have these constants defined when there's nowhere in the code to check that they're defined correctly. Maybe I could try to compute more things based on widths of windows? It happens in a few places already, but maybe more of them could be like `gWindows[...].width * TILE_WIDTH - howMuchPaddingThereIs` (where `howMuchPaddingThereIs` probably would be just an integer constant).

I get that those constants could be used in the tests, but I suppose what I'm saying is that it would be easy for somebody to change the UI and simply not see/update those constants, which would leave the tests effectively meaningless. Maybe for the tests which are computed based on windows I could make the window templates available globally, and then write computations involving those widths. (Although that might involve renaming some of them to have names which make sense globally, e.g. can't have a `gWindowTemplates`, it'd need to be `gPokemonStorageWindowTemplates` or w/e)

mrgriffin commented on gflib/text.c:
```diff
@@ -2012,3 +2084,33 @@ static void DecompressGlyph_Bold(u16 glyphId)
     gCurGlyph.height = 12;
 }
+
+static const s8 sNarrowerFontIds[] =
```
We don't have a `FONT_COUNT`, although ig I could introduce one? atm all the other arrays that involve fonts just use `[]` too.

AsparagusEduardo replied:
Weird, I thought we did. Maybe I had my cables crossed with something else else.

mrgriffin commented:
(Sorry about the force pushes, because there's many meaningful commits but some of them touch the same files I am doing the rebases locally rather than having a load of `!fixup` commits with conflicts that I'd have to resolve at the end)

AsparagusEduardo commented on gflib/window.c:
```diff
@@ -712,3 +712,13 @@ static u32 GetNumActiveWindowsOnBg8Bit(u32 bgId)
     return windowsNum;
 }
+
+u32 WindowWidthPx(u32 windowId)
+{
+    return gWindows[windowId].window.width * TILE_WIDTH;
+}
```
Is there a risk of OOB here?

mrgriffin replied:
Yeah, if you pass an invalid `windowId` then you'll read OOB. I'm not particularly worried about people doing that, because whatever code is asking for the width is about to use that same `windowId` to draw text. But ig we could put in a `if (windowId >= MAX_WINDOWS) return ?;`, although I'm not sure exactly what we'd want to return. `0`?

AsparagusEduardo replied:
Fair, I wouldn't know what to return either.

AsparagusEduardo commented on include/constants/global.h:
```diff
@@ -8,6 +8,9 @@
 #include "config/pokemon.h"
 #include "config/overworld.h"
 #include "config/decap.h"
+#if TESTING
+#include "config/test.h"
+#endif
```
Won't this do a redefinition warning thing because you're technically undefing them before they're defined?

mrgriffin replied:
I think they should all be defined by this point, because they come from the other `config/*.h` includes above this line :thinking:

`#undef` will work regardless of whether they're defined, but yeah for sure you'll get duplicate definition errors if a later header tries to define them.

AsparagusEduardo replied:
Right, though `POKEMON_NAME_LENGTH` isn't

mrgriffin replied:
Ah, damn! Good catch :) I'll move this to the end of the file

mrgriffin commented:
Addressed latest review comments :)

Please give me a chance to rebase before merging!

AsparagusEduardo commented:
> Addressed latest review comments :)
> 
> Please give me a chance to rebase before merging!

I was going to squash it anyway 🤔

mrgriffin commented:
> I was going to squash it anyway 🤔

The individual commits are potentially quite helpful—they're set up so that somebody could revert just the species UI changes, move UI changes, or item UI changes; while keeping the new fonts and the ability to turn on expanded species names and item names :)

mrgriffin commented:
Addressed the many bugs Edu found on Discord (thanks for being so thorough!!). Updated the screenshots to show the new things, and ready for re-review.

kittenchilly commented:
Since 12 character names isn't that much bigger than 10 characters, is there a case where we can't just extend the UI slightly?

mrgriffin commented:
> Since 12 character names isn't that much bigger than 10 characters, is there a case where we can't just extend the UI slightly?

This PR covers species names, move names, and item names. Item names in particular are not a slight extension of the UI.

The purpose of this PR was to pick something that doesn't require any UI changes, a) because it may clash with downstream users who have made changes to their UIs, and b) because that's the sort of change that attracts a lot of strong and conflicting opinions, which I can't be bothered to deal with.

ZnogyroP commented:
> Since 12 character names isn't that much bigger than 10 characters, is there a case where we can't just extend the UI slightly? Instead of using the font.

I know for a fact at least from my experience that at minimum, the Pokedex and the summary screen do not have enough space for 12 characters without significant UI changes. I think changing the UI is out of scope anyway, and if anyone does decide to do that they can just turn off the narrower fonts.

mrgriffin commented:
Oops, just need to update the tests to include the new things you discovered Edu! Should only take a couple of minutes though :)

mrgriffin commented:
There we go, should be all ready to go now :crossed_fingers:

mrgriffin commented:
Rebased onto `upcoming` and did official type names while I was at it (see updated screenshots). These will make Terastalization a little nicer, saying "Fighting" rather "Fight", for example.

AsparagusEduardo commented:
We're almost there xD

mrgriffin commented on src/battle_main.c:
```diff
@@ -295,6 +295,12 @@ const struct OamData gOamData_BattleSpritePlayerSide =
 static const s8 sCenterToCornerVecXs[8] ={-32, -16, -16, -32, -32};
 
+#if B_EXPANDED_TYPE_NAMES == TRUE
+#define HANDLE_EXPANDED_TYPE_NAME(_name, ...) _(DEFAULT(_name, __VA_ARGS__))
```
Oops!! :sweat_smile:

mrgriffin commented:
Okay, fixed the missing `HANDLE_EXPANDED_TYPE_NAME`s so that hopefully is the last change! :crossed_fingers:
